src/main/java/hudson/remoting/PingThread.java

```diff
@@ -106,31 +120,45 @@ public void run() {
 
     private void ping() throws IOException, InterruptedException {
         LOGGER.log(Level.FINE, "pinging {0}", channel.getName());
-        Future<?> f = channel.callAsync(new Ping());
+
+        int timeouts = 0;
         long start = System.currentTimeMillis();
 
-        long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout);
-        long remaining = end - System.nanoTime();
-
-        do {
-            LOGGER.log(Level.FINE, "waiting {0}s on {1}", new Object[] {TimeUnit.NANOSECONDS.toSeconds(remaining), channel.getName()});
-            try {
-                f.get(Math.max(1,remaining),TimeUnit.NANOSECONDS);
-                LOGGER.log(Level.FINE, "ping succeeded on {0}", channel.getName());
-                return;
-            } catch (ExecutionException e) {
-                if (e.getCause() instanceof RequestAbortedException)
-                    return; // connection has shut down orderly.
-                onDead(e);
-                return;
-            } catch (TimeoutException e) {
-                // get method waits "at most the amount specified in the timeout",
-                // so let's make sure that it really waited enough
+        while (true) {
```

Comment:
Seems like this would be better expressed as a `for` loop.

Reply:
Has been addressed in latest commit. In fact it was possible to entirely get rid of the inner loop.